user: Guard Linux-only -K options on FreeBSD (fixes #86252)

Jkhall81 opened the pull request:
SUMMARY
Fix handling of `uid_min`, `uid_max`, and `umask` options in `ansible.builtin.user` on FreeBSD.
FreeBSD’s `pw useradd` command does not support the Linux-specific `-K` option, causing failures when these parameters are set.
This change ensures these options are only applied when `useradd` is available (Linux), matching documentation stating they are ignored on non-Linux platforms.

Fixes: #86252

Signed-off-by: Jason Hall <jason.kei.hall@gmail.com>

ISSUE TYPE
Bugfix Pull Request

COMPONENT NAME
lib/ansible/modules/user.py

Jkhall81 committed:
user: Guard Linux-only -K options on FreeBSD (fixes #86252)

Akasurde commented on changelogs/fragments/freebsd-user-ignore-uidmin.yml:
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+  - "user module: FreeBSD backend now correctly ignores uid_min, uid_max, and umask,
```
```suggestion
  - "user: FreeBSD backend now correctly ignores uid_min, uid_max, and umask, instead of passing unsupported '-K' options to pw(8) (https://github.com/ansible/ansible/issues/86252)."
```

Akasurde commented on lib/ansible/modules/user.py:
```diff
@@ -1507,9 +1507,10 @@ def create_user(self):
-            if self.umask is not None:
-                cmd.append('-K')
-                cmd.append('UMASK=' + self.umask)
+            if self.module.get_bin_path('useradd', required=False):
```
Add a small comment saying why this is required.

Akasurde commented on lib/ansible/modules/user.py:
```diff
@@ -1526,13 +1527,14 @@ def create_user(self):
-        if self.uid_min is not None:
-            cmd.append('-K')
-            cmd.append('UID_MIN=' + str(self.uid_min))
+        if self.module.get_bin_path('useradd', required=False):
```
I would add an integration test for this change.

Jkhall81 committed:
user: Address review feedback for FreeBSD -K guard (issue #86252)

- Added explanatory comments above both Linux-only -K guards
- Updated changelog fragment per reviewer suggestion
- Added integration test to validate uid_min/uid_max/umask behavior on Linux

pkingstonxyz commented on test/integration/targets/user/tasks/uidmin_linux_behavior.yml:
```diff
@@ -0,0 +1,32 @@
+  register: user_result
+  become: true
+
+- name: Ensure module did not fail
```
I don't think this assert is necessary. If the module fails, then the test will simply fail without this assertion. If you'd like custom failed behavior on the task see how `test_create_user_min_max.yml` uses `failed_when`.

pkingstonxyz commented on test/integration/targets/user/tasks/uidmin_linux_behavior.yml:
```diff
@@ -0,0 +1,32 @@
+- name: Assert user exists
+  ansible.builtin.assert:
+    that:
+      - passwd_entry is not failed
```
I don't think this sufficiently tests the new behavior because it only checks if the user exists without validating any behavior around uid_min, uid_max, or umask. A brief sketch of a test that validates this might look something like:

```yml
- name: Create a user with uid_min and uid_max set to the same uid number

- assert:
  that:
    - the user creation didn't fail
    - the user's uid is NOT the specified uid number  # this checks that the bounds were ignored as the documentation states

- name: stat the user's home directory

- assert: ^ umask information from stat
```

Might be worth checking out `test_user_create_min_max.yml` and `test_umask.yml` for how this behavior is currently tested.

pkingstonxyz commented on test/integration/targets/user/tasks/uidmin_linux_behavior.yml:
This file needs to be `include_tasks`ed in the `user/tasks/main.yml` file

Jkhall81 committed:
user: Address review feedback for FreeBSD -k guard (issue #56252)

pkingstonxyz commented on test/integration/targets/user/tasks/main.yml:
```diff
@@ -34,3 +34,4 @@
     - ansible_facts.system == 'Linux'
     - ansible_distribution != 'Alpine'
 - import_tasks: ssh_keygen.yml
+- import_tasks: uidmin_linux_behavior.yml
```
Because all of the tasks in the file check if `ansible_os_family == 'FreeBSD'` we can just use a `when: ansile_facts.system == 'FreeBSD'` on the include block and then remove the rest of the `when` checks.

Jkhall81 replied:
I didn't know I could add that there.  This has been cleaned up.  Thanks.

Jkhall81 committed:
user: Clean up FreeBSD test conditions per review

pkingstonxyz commented on changelogs/fragments/freebsd-user-ignore-uidmin.yml:
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+  - "user: FreeBSD backend now correctly ignores uid_min, uid_max, and umask,
```
There's a small formatting issue here, see the docs for the correct format https://docs.ansible.com/projects/ansible/latest/community/development_process.html#changelog-fragments . For multiline strings, https://yaml-multiline.info/ might be useful here.
